### ISSUE TITLE
Simplify mapping iterables in toData and fromData implementations of common domain model classes

### ISSUE THREAD

**Issue by navibyte** [navibyte/dataflow#12]
See [sample domain model classes with JSON serialization](https://github.com/navibyte/dataflow/issues/10)

On this sample `PersonCollection` is defined as:
```dart
class PersonCollection {
  final Iterable<Person> persons;

  const PersonCollection({required this.persons});

  factory PersonCollection.fromData(DataArray data) => PersonCollection(
        persons: data.objects
            .map((element) => Person.fromData(element))
            .toList(growable: false),
      );

  DataArray toData() => DataArray.of(
        persons.map((person) => person.toData()).toList(growable: false),
      );
}
```

Mapping iterables in both directions in fromData and toData could be easier to implement?

Find choices and add support on `DataObject` and/or `DataArray`.

**Comment by navispatial** [comment]
DataArray with new static constructor:

```dart
  static DataArray from<T extends Object>(
          Iterable<T> source, Object Function(T) convert) =>
      DataArrayView._protected(
          source.map<Object>(convert).toList(growable: false));
```

And DataElement with new method (implemented by DataArray and DataArrayView etc.)

```dart
  /// Returns a list of [T] mapped from child data arrays using [map] function.
  ///
  /// An optional [limit] when provided limits the number of returned objects.
  ///
  /// The returned list is immutable.
  ///
  /// Other child objects (that cannot be represented as a data array) are
  /// omitted from an iterable.
  List<T> arraysToList<T extends Object>(T Function(DataArray array) map,
      {int? limit});
}
```

This allows simplification of our sample to:

```dart
class PersonCollection {
  final Iterable<Person> persons;

  const PersonCollection({required this.persons});

  static PersonCollection fromData(DataArray data) =>
      PersonCollection(persons: data.objectsToList(Person.fromData));

  DataArray toData() =>
      DataArray.from<Person>(persons, (person) => person.toData());
}
```

Also other similar changes and improvements...

**Comment by navispatial** [comment]
Implemented in [Release version 0.8.0](https://github.com/navibyte/dataflow/releases/tag/v0.8.0)
